arch/arm/core/irq_relay.S

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (c) 2018 Ding Tao
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file irq_relay.S
+ *
+ * @brief IRQ relay vector table and relay handler for Cortex-M0 SoC
+ *
+ * In Cortex-M0, the vector table address can not be changed. Once the
+ * vector table is occupied by bootloader, there will be no IRQ support
+ * in chainloaded image.
+ *
+ * This program will link into bootloader, once an interrupt is coming,
+ * bootloader can forward the interrupt to the chainloaded image. This
+ * will support DFU on Cortex-M0 plantform.
+ *
+ * Note: Currently support mcuboot only.
+ * */
+
+#include <board.h>
+#include <toolchain.h>
+#include <linker/sections.h>
+#include <drivers/system_timer.h>
+
+_ASM_FILE_PROLOGUE
+
+#if defined(CONFIG_IRQ_SOFT_RELAY)
+
+GDATA(_vector_table_pointer)
+GDATA(_main_stack)
+
+SECTION_FUNC(vector_relay_handler, __vector_relay_handler)
+	cpsid	i;
+
+	mrs		r0, ipsr;
+	lsls	r0, r0, $0x02;
+
+	ldr		r1, = _vector_table_pointer;
+	ldr		r1, [r1];
+	adds	r1, r1, r0;
+
```

Comment:
You are missing another
`ldr r1, [r1]` here to load the pointer from the vector table